File: Algorithms/Sorting/1_MergeSort/Python/mergeSort.py
def mergeSort(array):
    if len(array) <= 1:
        return array
    
    mid = len(array) // 2
    left = mergeSort(array[:mid])
    right = mergeSort(array[mid:])
    
    return merge(left, right)


def merge(listLeft, listRight):
    result = []
    i = 0
    j = 0
    while i < len(listLeft) and j < len(listRight):
        if listLeft[i] < listRight[j]:
            result.append(listLeft[i])
            i += 1
        else:
            result.append(listRight[j])
            j += 1

    while i < len(listLeft):
        result.append(listLeft[i])
        i += 1

    while j < len(listRight):
        result.append(listRight[j])
        j += 1

    return result


print(mergeSort([3, 5, 1, 6]))

File: Algorithms/Sorting/1_MergeSort/Rust/merge_sort.rs
fn main() {
    let arr = vec![12, 11, 13, 5, 6, 7];
    let sorted_arr = merge_sort(arr);

    print!("Sorted array: ");
    for i in 0..sorted_arr.len() {
        print!("{}, ", sorted_arr[i]);
    }
}

fn merge_sort(arr: Vec<i32>) -> Vec<i32> {
    if arr.len() <= 1 {
        return arr;
    }

    let midle = arr.len() / 2;
    let mut left = Vec::new();
    let mut right = Vec::new();

    for i in 0..midle {
        left.push(arr[i]);
    }

    for i in midle..arr.len() {
        right.push(arr[i]);
    }

    left = merge_sort(left);
    right = merge_sort(right);
    return merge(left, right);
}

fn merge(left: Vec<i32>, right: Vec<i32>) -> Vec<i32> {
    let mut result = Vec::new();
    let mut left_index = 0;
    let mut right_index = 0;

    while left_index < left.len() && right_index < right.len() {
        if left[left_index] < right[right_index] {
            result.push(left[left_index]);
            left_index += 1;
        } else {
            result.push(right[right_index]);
            right_index += 1;
        }
    }

    while left_index < left.len() {
        result.push(left[left_index]);
        left_index += 1;
    }

    while right_index < right.len() {
        result.push(right[right_index]);
        right_index += 1;
    }

    return result;
}
 
